flink-runtime/src/main/java/org/apache/flink/runtime/util/EnvironmentInformation.java

```diff
@@ -137,7 +137,13 @@ public static long getSizeOfFreeHeapMemoryWithDefrag() {
 	 */
 	public static long getSizeOfFreeHeapMemory() {
 		Runtime r = Runtime.getRuntime();
-		return r.maxMemory() - r.totalMemory() + r.freeMemory();
+		long maxMemory = r.maxMemory();
+		if (maxMemory == Long.MAX_VALUE) {
+			// workaround for some JVM versions
+			return r.freeMemory();
```

Comment:
OK, didn't see it yesterday. Maybe we can play around with [MemoryUsage](https://docs.oracle.com/javase/7/docs/api/java/lang/management/MemoryUsage.html).